include/cocaine/framework/receiver.hpp

```diff
@@ -0,0 +1,324 @@
+        }
+
+        result_type operator()(error_type& error) const {
+            throw cocaine_error(error);
+        }
+
+        result_type operator()(choke_type&) const {
+            return boost::none;
+        }
+    };
+};
+
+template<class T>
+class terminator {
+    typedef std::function<bool()> function_type;
+
+public:
+    static std::vector<function_type> generate() {
+        std::vector<function_type> result;
+        boost::mpl::for_each<typename io::protocol<T>::messages>(terminator<T>(result));
+        return result;
+    }
+
+    template<class U>
+    void operator()(const U&) {
+        unpackers.emplace_back(&terminator::unpacker<U>::unpack);
+    }
+
+private:
+    std::vector<function_type>& unpackers;
+
+    terminator(std::vector<function_type>& unpackers) :
+        unpackers(unpackers)
+    {}
+
+public:
+    template<typename U>
+    struct unpacker {
+        static
+        bool
+        unpack() {
+            return std::is_same<typename io::event_traits<U>::dispatch_type, void>::value;
+        }
+    };
+};
+
+template<class T, class Session>
+class receiver {
+    typedef typename variant_of<T>::type variant_type;
+    typedef typename variant_type::types variant_typelist;
+
+    typedef std::function<variant_type(const msgpack::object&)> unpacker_type;
+    typedef std::function<bool()> terminator_type;
+
+    static const std::vector<unpacker_type> visitors;
+    static const std::vector<terminator_type> terminators;
+
+    std::shared_ptr<basic_receiver_t<Session>> d;
+
+public:
+    receiver(std::shared_ptr<basic_receiver_t<Session>> d) :
+        d(d)
+    {}
+
+    // Intentionally deleted.
+    receiver(const receiver&) = default;
+    receiver(receiver&&) = default;
+
+    receiver& operator=(const receiver&) = default;
+    receiver& operator=(receiver&&) = default;
+
+    /*!
+     * Returns a future, which contains either the next protocol message or the exception.
+     *
+     * \note this method automatically revokes the channel when reached a leaf in the dispatch graph.
+     *
+     * \warning this receiver may be invalidated after this call.
+     */
+    typename task<typename receiver_traits<T>::result_type>::future_type
+    recv() {
+        return d->recv()
+            .then(std::bind(&receiver<T, Session>::convert, std::placeholders::_1, d));
+    }
+
+private:
+    static
+    typename receiver_traits<T>::result_type
+    convert(typename task<decoded_message>::future_type& f, std::shared_ptr<basic_receiver_t<Session>> d) {
+        const decoded_message message = f.get();
+        const std::uint64_t id = message.type();
+
+        // Some ancient boost::mpl versions return sized integer instead of unsized one.
+        if (id >= static_cast<size_t>(boost::mpl::size<variant_typelist>::value)) {
+            // TODO: What to do? Notify the user, I think.
+            throw std::runtime_error("convert message type check: not implemented yet");
```

Comment:
It may be a completely irrelevant protocol, not just newer version of the correct one. So this message may be confusing.